File: 0001-0500/0402. Remove K Digits.go
package main
// Medium
// Time:O(n), Space:O(n)
// https://leetcode.cn/problems/remove-k-digits/

import "strings"
func removeKdigits(num string, k int) string {
    stack := []byte{}

    for i := 0; i < len(num); i++ {
        for len(stack) > 0 && k > 0 && stack[len(stack) - 1] > num[i] {
            stack = stack[:len(stack) - 1]
            k--
        }
        stack = append(stack, num[i])
    }

    stack = stack[:len(stack) - k]
    // remove prefix-0
    res := strings.TrimLeft(string(stack), "0")

    if res == "" {
        return "0"
    }
    return res
}

File: 0001-0500/0402. Remove K Digits.py
# Time:O(n), Space:O(n)

class Solution:
    def removeKdigits(self, num: str, k: int) -> str:
        stack = []

        for d in num:
            while stack and k > 0 and stack[-1] > d:
                stack.pop()
                k -= 1

            stack.append(d)

        while k > 0:
            stack.pop()
            k -= 1

        res = "".join(stack).lstrip("0")

        return res if res else "0"
